spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractMessageListenerContainer.java

```diff
@@ -399,16 +399,30 @@ public void checkGroupId() {
 
 	@Override
 	public final void stop() {
+		stop(true);
+	}
+
+	/**
+	 * Stop the container.
+	 * @param wait wait for the listener to terminate.
+	 * @since 2.3.8
+	 */
+	public final void stop(boolean wait) {
 		synchronized (this.lifecycleMonitor) {
 			if (isRunning()) {
-				final CountDownLatch latch = new CountDownLatch(1);
-				doStop(latch::countDown);
-				try {
-					latch.await(this.containerProperties.getShutdownTimeout(), TimeUnit.MILLISECONDS); // NOSONAR
-					publishContainerStoppedEvent();
+				if (wait) {
+					final CountDownLatch latch = new CountDownLatch(1);
+					doStop(latch::countDown);
+					try {
+						latch.await(this.containerProperties.getShutdownTimeout(), TimeUnit.MILLISECONDS); // NOSONAR
+						publishContainerStoppedEvent();
+					}
+					catch (@SuppressWarnings("unused") InterruptedException e) {
+						Thread.currentThread().interrupt();
+					}
 				}
-				catch (@SuppressWarnings("unused") InterruptedException e) {
-					Thread.currentThread().interrupt();
+				else {
+					doStop(() -> { });
```

Comment:
When `publishContainerStoppedEvent();` is emitted in this case, when we don't wait for container to be stopped?

On the other hand what is the point of such an event if we still wait for container to be stopped?..

Reply:
Sorry; I don't understand; the no-wait stop is when the container stops itself; previously it deadlocked for the shutdown timeout.

What does this have to do with event publishing?

Reply:
OK. So, previously we always waited for the stop and then emitted an event.
Right now we have a `wait` condition and emit an event only when it is `true`.
In case of `false` we have only this non-blocking `doStop()`.
I don't mind about "container stops itself" logic, but I try to understand why we don't emit an event when we don't wait for stop.

Hope this is much cleaner now.

Reply:
Sorry; missed that; fixed.

Reply:
Docs are OK, but I can't merge without an explanation over here.

Thanks for understanding.